go/engine/pgp_push_private.go

```diff
@@ -0,0 +1,218 @@
+// Copyright 2015 Keybase, Inc. All rights reserved. Use of
+// this source code is governed by the included BSD license.
+
+package engine
+
+import (
+	"errors"
+	"fmt"
+	"github.com/keybase/client/go/libkb"
+	keybase1 "github.com/keybase/client/go/protocol/keybase1"
+	"github.com/keybase/go-framed-msgpack-rpc/rpc"
+)
+
+type PGPPushPrivate struct {
+	arg keybase1.PGPPushPrivateArg
+}
+
+func (e *PGPPushPrivate) Name() string {
+	return "PGPPushPrivate"
+}
+
+func (e *PGPPushPrivate) Prereqs() Prereqs {
+	return Prereqs{}
+}
+
+func (e *PGPPushPrivate) RequiredUIs() []libkb.UIKind {
+	return []libkb.UIKind{}
+}
+
+func (e *PGPPushPrivate) SubConsumers() []libkb.UIConsumer {
+	return []libkb.UIConsumer{}
+}
+
+func NewPGPPushPrivate(arg keybase1.PGPPushPrivateArg) *PGPPushPrivate {
+	return &PGPPushPrivate{arg}
+}
+
+func getCurrentUserPGPKeys(m libkb.MetaContext) ([]libkb.PGPFingerprint, error) {
+	uid := m.CurrentUID()
+	if uid.IsNil() {
+		return nil, libkb.NewLoginRequiredError("for push/pull of PGP private keys to KBFS")
+	}
+	upk, _, err := m.G().GetUPAKLoader().LoadV2(libkb.NewLoadUserArgWithMetaContext(m).WithUID(uid))
+	if err != nil {
+		return nil, err
+	}
+	var res []libkb.PGPFingerprint
+	for _, key := range upk.Current.PGPKeys {
+		if key.Base.Revocation != nil {
+			continue
+		}
+		res = append(res, libkb.PGPFingerprint(key.Fingerprint))
+	}
+	if len(res) == 0 {
+		return nil, errors.New("The --all flag only works if you have PGP keys linked to your Keybase account")
+	}
+	return res, nil
+}
+
+func getPrivateFingerprints(m libkb.MetaContext, fps []keybase1.PGPFingerprint) ([]libkb.PGPFingerprint, error) {
+	if len(fps) == 0 {
+		return getCurrentUserPGPKeys(m)
+	}
+
+	var ret []libkb.PGPFingerprint
+	for _, fp := range fps {
+		ret = append(ret, libkb.ImportPGPFingerprint(fp))
+
+	}
+	return ret, nil
+}
+
+func simpleFSClient(m libkb.MetaContext) (*keybase1.SimpleFSClient, error) {
+	xp := m.G().ConnectionManager.LookupByClientType(keybase1.ClientType_KBFS)
+	if xp == nil {
+		return nil, libkb.KBFSNotRunningError{}
+	}
+	return &keybase1.SimpleFSClient{
+		Cli: rpc.NewClient(xp, libkb.NewContextifiedErrorUnwrapper(m.G()), nil),
+	}, nil
+}
+
+func (e *PGPPushPrivate) mkdir(m libkb.MetaContext, fs *keybase1.SimpleFSClient, path string) (err error) {
+	opid, err := fs.SimpleFSMakeOpid(m.Ctx())
+	if err != nil {
+		return err
+	}
+	defer fs.SimpleFSClose(m.Ctx(), opid)
```

Comment:
Same here, making a directory doesn't leave a handle hanging around that needs to be cleaned up.  (Yes, this interface kind of sucks.)

Reply:
Should I `Wait`?

Reply:
BTw, for this, I'm doing what's in `cmd_simplefs_mkdir.go`

Reply:
You don't need to `Wait` either; it will error too.  My earlier statement wasn't totally accurate: you only need to wait when using an opID to calls that are asynchronous.  Making a directory is not one of these (though I realize that's totally not obvious.)

I guess it's a (harmless) bug that the `fs mkdir` does a close.